### PR TITLE
virtqueue: Add virtqueue_set_shmem_io & remove passing shmem_io in virtqueue_create

### PR DESCRIPTION
**Pull request by galak**
Pull in changes to the virtqueue from staging branch. This is based on PR #50

### CODE REVIEW

File: lib/rpmsg/remote_device.c [hunk 1]
```diff
@@ -428,11 +428,17 @@ int rpmsg_rdev_create_virtqueues(struct virtio_device *dev, int flags, int nvqs,
 				vring_table[idx].align));
 		}
 
+		vqs[idx] = virtqueue_allocate(ring_info.num_descs);
+		if (vqs[idx] == NULL) {
+			return (ERROR_NO_MEM);
```
Comment (skordal) [inline lib/rpmsg/remote_device.c]: I think this should be `return RPMSG_ERR_NO_MEM;` to be consistent with the namespace.

File: lib/virtio/virtqueue.c [hunk 2]
```diff
@@ -44,49 +43,35 @@ int virtqueue_create(struct virtio_device *virt_dev, unsigned short id,
...
-		/* Initialize vring control block in virtqueue. */
-		vq_ring_init(vq, (void *)ring->vaddr, ring->align);
-
-		/* Disable callbacks - will be enabled by the application
-		 * once initialization is completed.
-		 */
-		virtqueue_disable_cb(vq);
-
-		*v_queue = vq;
+	if (vq == NULL) {
```
Comment (skordal) [inline lib/virtio/virtqueue.c]: This could be changed to `VQ_PARAM_CHK(vq == NULL, status, ERROR_NO_MEM);`

File: lib/virtio/virtqueue.c [hunk 3]
```diff
@@ -44,49 +43,35 @@ int virtqueue_create(struct virtio_device *virt_dev, unsigned short id,
...
+		return (ERROR_NO_MEM);
 	}
 
+	vq->vq_dev = virt_dev;
```
Comment (skordal) [inline lib/virtio/virtqueue.c]: The remaining code in this function (all the way down to line 74) has to be within an `if (status == VQUEUE_SUCCESS)` block, due to the way `VQ_PARAM_CHK` is defined.

File: lib/rpmsg/remote_device.c (hunk 1 above)
Comment (galak) [reply lib/rpmsg/remote_device.c]: Thanks will fix.

File: lib/virtio/virtqueue.c (hunk 2 above)
Comment (galak) [reply lib/virtio/virtqueue.c]: Will change.

File: lib/virtio/virtqueue.c (hunk 3 above)
Comment (galak) [reply lib/virtio/virtqueue.c]: Will fix.

**Comment by galak** [comment]
Closing in favor of a new PR with all vq changes.

**Pull request closed without merging**